src/PJ_pipeline.c

```diff
@@ -0,0 +1,603 @@
+* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+* DEALINGS IN THE SOFTWARE.
+*
+********************************************************************************/
+
+#define PJ_LIB__
+#include <proj.h>
+#include <projects.h>
+
+#include <assert.h>
+#include <stddef.h>
+#include <errno.h>
+PROJ_HEAD(pipeline,         "Transformation pipeline manager");
+
+/* Projection specific elements for the PJ object */
+#define PIPELINE_STACK_SIZE 100
+struct pj_opaque {
+    int reversible;
+    int steps;
+    int depth;
+    int verbose;
+    int *reverse_step;
+    int *omit_forward;
+    int *omit_inverse;
+    PJ_OBS stack[PIPELINE_STACK_SIZE];
+    PJ **pipeline;
+};
+
+
+static PJ_OBS pipeline_forward_obs (PJ_OBS, PJ *P);
+static PJ_OBS pipeline_reverse_obs (PJ_OBS, PJ *P);
+static XYZ    pipeline_forward_3d (LPZ lpz, PJ *P);
+static LPZ    pipeline_reverse_3d (XYZ xyz, PJ *P);
+static XY     pipeline_forward (LP lpz, PJ *P);
+static LP     pipeline_reverse (XY xyz, PJ *P);
+
+
+
+/********************************************************************
+
+                  ISOMORPHIC TRANSFORMATIONS
+
+*********************************************************************
+
+    In this context, an isomorphic transformation is a proj PJ
+    object returning the same kind of coordinates that it
+    receives, i.e. a transformation from angular to angular or
+    linear to linear coordinates.
+
+    The degrees-to-radians operation is an example of the former,
+    while the latter is typical for most of the datum shift
+    operations used in geodesy, e.g. the Helmert 7-parameter shift.
+
+    Isomorphic transformations trips up the pj_inv/pj_fwd
+    functions, which try to check input sanity and scale output to
+    internal proj units, under the assumption that if input is of
+    angular type, then output must be of linear (or vice versa).
+
+    Hence, to avoid having pj_inv/pj_fwd stomping on output (or
+    choking on input), we need a way to tell them that they should
+    accept whatever is being handed to them.
+
+    The P->left and P->right elements indicate isomorphism.
+
+    For classic proj style projections, they both have the
+    value PJ_IO_UNITS_CLASSIC (default initialization), indicating
+    that the forward driver expects angular input coordinates, and
+    provides linear output coordinates.
+
+    Newer projections may set P->left and P->right to either
+    PJ_IO_UNITS_METERS, PJ_IO_UNITS_RADIANS or PJ_IO_UNITS_ANY,
+    to indicate their I/O style.
+
+    For the forward driver, left indicates input coordinate
+    type, while right indicates output coordinate type.
+
+    For the inverse driver, left indicates output coordinate
+    type, while right indicates input coordinate type.
+
+*********************************************************************/
+
+
+
+
+static PJ_OBS pipeline_forward_obs (PJ_OBS point, PJ *P) {
+    int i, first_step, last_step, incr;
+
+    first_step = 1;
```

Comment:
It's actually mostly for historical reasons (in the previous version of the pipeline stuff, P[0] was the pipeline driver itself, and a sentinel was needed in the far end of the pipeline, because P was an array of PJs, not of PJ *.

I intend to clean this up at a later date, but I need to go on with some material actually using the pipeline functionality in order to stress test the implementation